overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,231 @@
+ */
+export class BucketEventSubscription extends lambda.EventSubscription {
+    public readonly bucket: pulumi.Output<Bucket>;
+
+    public constructor(
+        name: string, bucket: Bucket, func: lambda.Function,
+        args: BucketEventSubscriptionArgs, opts?: pulumi.ResourceOptions) {
+
+        super("aws-serverless:bucket:BucketEventSubscription", name, func, { bucket: bucket }, opts);
+
+        const permission = new lambda.Permission(name, {
+            function: func,
+            action: "lambda:InvokeFunction",
+            principal: "s3.amazonaws.com",
+            sourceArn: bucket.id.apply(bucketName => `arn:aws:s3:::${bucketName}`),
+        }, { parent: this });
+
+        this.permission = permission;
+
+        // We must create only a single BucketNotification per Bucket per AWS API limitations.  See
+        // https://github.com/terraform-providers/terraform-provider-aws/issues/1715.  So we push
+        // the subscription information here, and then actually create the BucketNotification if
+        // needed on process `beforeExit`.
+        let subscriptions = bucketSubscriptionInfos.get(bucket);
+        if (!subscriptions) {
+            subscriptions = [];
+            bucketSubscriptionInfos.set(bucket, subscriptions);
+        }
+
+        subscriptions.push({
+            name: name,
+            events: args.events,
+            filterPrefix: args.filterPrefix,
+            filterSuffix: args.filterSuffix,
+            lambdaFunctionArn: func.arn,
+            permission: permission,
+        });
+    }
+}
+
+process.on("beforeExit", () => {
+    const copy = bucketSubscriptionInfos;
+
+    // Since we are generating more work on the event loop, we will casue `beforeExit` to be invoked again.
+    // Make sure to clear out eh pending subscrpitions array so that we don't try to apply them again.
+    bucketSubscriptionInfos = new Map();
+
+    for (const [bucket, subscriptions] of copy) {
+        const permissions = subscriptions.map(s => s.permission);
+        const _ = new BucketNotification(subscriptions[0].name, {
+            bucket: bucket.id,
+            lambdaFunctions: subscriptions.map(subscription => ({
+                events: subscription.events,
+                filterPrefix: subscription.filterPrefix,
+                filterSuffix: subscription.filterSuffix,
+                lambdaFunctionArn: subscription.lambdaFunctionArn,
+            })),
+        }, { parent: bucket, dependsOn: permissions });
+    }
+});
+
+// Mixin event handling functionality onto Bucket.
+declare module "./bucket" {
+    interface Bucket {
+        onObjectCreated(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectCreatedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+
+        onObjectRemoved(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectRemovedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
```

Comment:
the above names are an opinionated change than those cloud/serverless used to use.  Instead of onPut/onDelete, we use the names that AWS uses for these operations.  i.e. when you register for these events you use the "s3:ObjectCreated" tag.  

This fits into our view that @pulumi/aws should be feel very aws native.

Reply:
I agree 100% with this.